lib/pos_blockchain.go

```diff
@@ -105,6 +102,20 @@ func (bc *Blockchain) processBlockPoS(block *MsgDeSoBlock, currentView uint64, v
 	return appliedNewTip, false, nil, nil
 }
 
+func (bc *Blockchain) processOrphanBlockPoS(block *MsgDeSoBlock) error {
+	// All blocks should pass the basic integrity validations, which ensure the block
+	// is not malformed. If the block is malformed, we should store it as ValidateFailed.
+	if err := bc.isProperlyFormedBlockPoS(block); err != nil {
+		if _, innerErr := bc.storeValidateFailedBlockInBlockIndex(block); innerErr != nil {
+			return errors.Wrapf(innerErr, "processBlockPoS: Problem adding validate failed block to block index: %v", err)
+		}
+		return nil
+	}
+	// Add to blockIndexByHash with status STORED only as we are not sure if it's valid yet.
```

Comment:
It'll be good to validate the block's proposer here too. As long as the block belong somewhere within the current or next epoch, it's possible to validate the proposer. Possibly the validator set too ideally.

I'd go as far as outright rejecting (erroring and not storing) blocks that aren't signed by the expected block proposer and with a QC from the expected validator set. We know that such blocks are outright bad and likely malicious. We won't even want to allocate disk space to store them so a malicious proposer doesn't try to fill up our disk with bad blocks

Reply:
It's a bit difficult to validate the QC and the block proposer as we currently create a utxo view based on the parent hash of this block. If we don't have the parent, we can't build the view and it makes figuring out the validator set and leader schedule tricky. Perhaps we can make some new helper functions that look up snapshots based on block heights to get around this issue.